arelle/plugin/validate/ROS/rules/ros.py

```diff
@@ -338,3 +338,30 @@ def rule_ros18(
             code='ROS.18',
             message=_("'%(conceptLn)s' cannot have a negative value.")
         )
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+)
+def rule_ros19(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> None:
+    """
+    ROS: Rule 19:PrincipalCurrencyUsedInBusinessReport must exist and its value must match the name of the unit used for all monetary facts.
+    """
+    monetaryFacts = val.modelXbrl.factsByDatatype(False, qnXbrliMonetaryItemType)
+    monetaryUnits = set(fact.unit.value for fact in monetaryFacts)
+    pricipalCurrencyFacts = val.modelXbrl.factsByLocalName.get(PRINCIPAL_CURRENCY, set())
+    principalCurrencyValues = set(fact.text for fact in pricipalCurrencyFacts)
+    if monetaryUnits != principalCurrencyValues:
```

Comment:
I think this would pass if there were two `PrincipalCurrencyUsedInBusinessReport` values and both the corresponding units were used in the report. If that wouldn't otherwise cause an error, we should probably capture that as an error here.